### ISSUE TITLE
"$tile-state" filters on background layer type

### ISSUE THREAD

**Issue by lucaswoj** [mapbox/mapbox-gl-style-spec#624]
## Requirements

 - designers have the ability to hide the base land color until all tiles at that coordinate have loaded so that a loading pattern (grid?) beneath the map can show through
 - designers have the ability to show a pattern over tiles that are still loading
 - designers have the ability to show a pattern over tiles whose requests have errored
 - style spec layer types map exclusively to shape types in semantics and names

## Design Alternatives

 - Add a layer type which renders a pattern over portions of the map at which all tiles have loaded https://github.com/mapbox/mapbox-gl-style-spec/issues/219#issuecomment-250873958
 - Add a layer type which renders a pattern over portions of the map at which all tiles are in a specified state (loaded, errored, loading) https://github.com/mapbox/mapbox-gl-js/issues/3280
 - Allow `$tile-state` `filter`s on `background` layer type

## Design

Allow `background` layers to have tile state filters.

```
{
    "type": "background",
    "filter": ["==", "$tile-state", "loaded"],
    "paint" {
        "background-color": "red"
    }
}
```

This keeps the mapping between style spec layer types and shape types, pushes us towards a unified expression syntax, and meets all the requirements stated above ☝️ 

(Note that this does not preclude renaming the `background` layer type to `cover` or `fill`. I think that's a great & orthogonal idea.)

### Concepts

Filters are a concept familiar to our users. The idea of using a filter on a tile rather than a feature is a new precedent. This new precedent fits with our plans of building a unified arbitrary expression syntax with a library of useful primitives.

**Comment by jfirebaugh** [comment]
My initial reaction is that using `filter` for this feels like an awkward shoehorning. A filter is a function applied to a _feature_, producing a boolean result indicating whether that feature should be included in the tile's layout. What we have here is something different: a function applied to a _source tile_, producing a boolean result indicating whether the tile should be rendered. Generally, different things should look different and have different names.

Some concrete questions about this proposal:

* Currently, `background` layers are unassociated with any source, and always rendered. Would we start requiring a `source` property for all `background` layers? Or only those that have a `filter`? (Or are you imagining that `source` remains prohibited and `["==", "$tile-state", "loaded"]` evaluated to false iff there's at least one source tile that isn't loaded? If so, imagine the case where there is a basemap source and an overlay source, and only the basemap tile is loaded. Its data would be rendered, but the ground/water background wouldn't. I don't think that's the desired behavior.)
* What happens if a `background` filter contains a key other than `"$tile-state"`?
* Is `$tile-state` allowed in filters of other layer types?

> designers have the ability to hide the base land color until all tiles at that coordinate have loaded so that a loading pattern (grid?) beneath the map can show through

I'd rephrase this requirement as: designers have the ability to permit a background loading pattern -- distinct from the background land or water color -- to show through portions of the map where no data has been loaded. (Key difference: no "all tiles" requirement. I don't think that's a real requirement, and in any case none of the proposed solutions satisfy it.)

**Comment by lucaswoj** [comment]
> My initial reaction is that using `filter` for this feels like an awkward shoehorning. A filter is a function applied to a _feature_, producing a boolean result indicating whether that feature should be included in the tile's layout. What we have here is something different: a function applied to a _source tile_, producing a boolean result indicating whether the tile should be rendered. Generally, different things should look different and have different names.

So far we've only provided facilities for filtering features. The concept of a "filter" is more general, however. For example, Javascript offers an `Array#filter` method which can operate on any type of data. 

Alternately, you might look at this as a proposal to define a "background" layer as being comprised of "tile features." 

> Currently, `background` layers are unassociated with any source, and always rendered. Would we start requiring a `source` property for all `background` layers? Or only those that have a `filter`? (Or are you imagining that `source` remains prohibited and `["==", "$tile-state", "loaded"]` evaluated to false iff there's at least one source tile that isn't loaded? If so, imagine the case where there is a basemap source and an overlay source, and only the basemap tile is loaded. Its data would be rendered, but the ground/water background wouldn't. I don't think that's the desired behavior.)

I'm still experimenting with each of these options to determine what feels most natural to use & implement. These questions are mostly orthogonal to the syntax designs being contrasted in this RFC. 

> What happens if a `background` filter contains a key other than `"$tile-state"`?

In the near term, this will cause a runtime error. In the long term, I see us unifying & expanding the capabilities of filters, functions, and cluster aggregators. Some expression primitives will be available in all cases (e.g. the `==` operator) and some may be available only in specific cases.

>  Is `$tile-state` allowed in filters of other layer types?

I see no conceptual reason why we couldn't expose `$tile-state` to other layer types. As we work to to unify & expand the capabilities of filters, functions, and cluster aggregators we may choose to expose `$tile-state` to other layer types. There is no strong incentive to implement this feature in the near term.

**Comment by jfirebaugh** [comment]
I outlined my preferred solution in #626.

**Comment by 1ec5** [comment]
I share the concern that @jfirebaugh expresses in https://github.com/mapbox/mapbox-gl-style-spec/issues/624#issuecomment-267490201 about shoehorning the filter concept into this problem. On the other hand, if there’s the possibility in the future of valid variables beyond `$tile-state`, I can see how having access to operators like `any` and `in` may be useful.

We could mitigate these concerns by distinguishing `Filter` objects from the `filter` property on layer objects. The `Filter` type is a generally useful construct for expressing conditions, whereas the `filter` property is useful specifically for evaluating whether a given feature should be included in the layer. The `Filter` type defines its structure and operators, whereas the `filter` property defines semantics – allowed variables such as `$id` and source layer variables. This is no different than how [`Array`](https://www.mapbox.com/mapbox-gl-style-spec/#types-array) is always an array, but `line-dasharray` defines the array’s contents and how they’re interpreted.

`background` layers could instead have a `tile-filter` property (tentatively named) that is set to a `Filter` object. Unlike the `filter` property, which remains unavailable to `background` layers, the `tile-filter` property allows only the `$tile-state` variable, with more variables becoming available in the future.

**Comment by lucaswoj** [comment]
> On the other hand, if there’s the possibility in the future of valid variables beyond `$tile-state`, I can see how having access to operators like `any` and `in` may be useful.

I expect we will need to include multiple valid variables in the initial implementation in order to meet all use cases. 

One potential design has us providing one variable per source:

```
"filter": ["any", ["==", "$mapbox-tile-state", "loading"], ["==", "$1ec5-tile-state", "loading"]]
```

We may also need to provide some differentiation between a tile that's loading but covered by overzoomed data and a tile that is loading and not covered by overzoomed data. 

----

> `background` layers could instead have a `tile-filter` property 

I like this idea 😄

**Comment by lucaswoj** [comment]
These are the 4 use cases I anticipate this feature needing to address. I'm working with clients and internal teams to see if any of these are unrealistic or if there are additional use cases we should consider. 

 1. Display a pattern over failed tiles from a particular source
 2. Display a pattern over loading or overzoomed tiles from a particular source
 3. Hide background tiles under loading (but not overzoomed) tiles from a particular source
 4. Hide background tiles under loading (but not overzoomed) tiles from all sources

**Comment by lucaswoj** [comment]
Based on some conversations with customers and internal teams, I think #626 is the best choice for now 👍
